interfaces/builtin/dbus.go

```diff
@@ -0,0 +1,430 @@
+
+# allow connected snaps to all interfaces via ###DBUS_PATH### (eg,
+# org.freedesktop.*, org.gtk.Application, etc) to allow full integration with
+# connected snaps.
+dbus (receive, send)
+    bus=###DBUS_BUS###
+    path=###DBUS_PATH###
+    peer=(label=###PLUG_SECURITY_TAGS###),
+`
+
+const dbusConnectedPlugAppArmor = `
+#include <abstractions/###DBUS_ABSTRACTION###>
+
+# allow snaps to introspect the slot servive. This allows us to introspect
+# all DBus interfaces of the service (but not use them).
+dbus (send)
+    bus=###DBUS_BUS###
+    interface=org.freedesktop.DBus.Introspectable
+    member=Introspect
+    peer=(label=###SLOT_SECURITY_TAGS###),
+
+# allow connected snaps to ###DBUS_NAME###
+dbus (receive, send)
+    bus=###DBUS_BUS###
+    peer=(name=###DBUS_NAME###, label=###SLOT_SECURITY_TAGS###),
+# For KDE applications, also support alternation since they use org.kde.foo-PID
+# as their 'well-known' name. snapd does not allow ###DBUS_NAME### to end with
+# '-[0-9]+', so this is ok.
+dbus (receive, send)
+    bus=###DBUS_BUS###
+    peer=(name=###DBUS_NAME###-[1-9]{,[0-9]}{,[0-9]}{,[0-9]}{,[0-9]}{,[0-9]}, label=###SLOT_SECURITY_TAGS###),
+
+# allow connected snaps to all paths via ###DBUS_INTERFACE### to allow full
+# integration with connected snaps.
+dbus (receive, send)
+    bus=###DBUS_BUS###
+    interface=###DBUS_INTERFACE###
+    peer=(label=###SLOT_SECURITY_TAGS###),
+
+# allow connected snaps to all interfaces via ###DBUS_PATH### (eg,
+# org.freedesktop.*, org.gtk.Application, etc) to allow full integration with
+# connected snaps.
+dbus (receive, send)
+    bus=###DBUS_BUS###
+    path=###DBUS_PATH###
+    peer=(label=###SLOT_SECURITY_TAGS###),
+`
+
+const dbusConnectedPlugSecComp = `
+getsockname
+recvmsg
+sendmsg
+sendto
+`
+
+type DbusInterface struct{}
+
+func (iface *DbusInterface) Name() string {
+	return "dbus"
+}
+
+// Obtain yaml-specified bus well-known name
+func (iface *DbusInterface) getAttribs(attribs map[string]interface{}) (string, string, error) {
+	// bus attribute
+	bus, ok := attribs["bus"].(string)
+	if !ok {
+		return "", "", fmt.Errorf("cannot find attribute 'bus'")
+	}
+
+	if bus != "session" && bus != "system" {
+		return "", "", fmt.Errorf("bus '%s' must be one of 'session' or 'system'", bus)
+	}
+
+	// name attribute
+	name, ok := attribs["name"].(string)
+	if !ok {
+		return "", "", fmt.Errorf("cannot find attribute 'name'")
+	}
+
+	err := interfaces.ValidateDBusBusName(name)
+	if err != nil {
+		return "", "", err
+	}
+
+	// snapd has AppArmor rules (see above) allowing binds to busName-PID
+	// so to avoid overlap with different snaps (eg, busName running as PID
+	// 123 and busName-123), don't allow busName to end with -PID. If that
+	// rule is removed, this limitation can be lifted.
+	invalidSnappyBusName := regexp.MustCompile("-[0-9]+$")
+	if invalidSnappyBusName.MatchString(name) {
+		return "", "", fmt.Errorf("DBus bus name must not end with -NUMBER")
+	}
+
+	if bus == "" {
```

Comment:
I'm confused by these checks, it doesn't seem that bus == "" would pass the previous check being equal session or system,  same for name == "" and ValidateDBusName ?